graphics/lvgl/port/lv_port_fbdev.c

```diff
@@ -0,0 +1,755 @@
+                  *cur_pos = lv_color_to32(*color_p);
+                  cur_pos++;
+                  color_p++;
+                }
+            }
+        }
+        break;
+      case 16:
+        {
+          FAR uint16_t *fbp = fbdev_obj->act_buffer;
+          for (y = y1; y <= y2; y++)
+            {
+              FAR uint16_t *cur_pos = fbp + (y * xres) + x1;
+              for (x = x1; x <= x2; x++)
+                {
+                  *cur_pos = lv_color_to16(*color_p);
+                  cur_pos++;
+                  color_p++;
+                }
+            }
+        }
+        break;
+      case 8:
+        {
+          FAR uint8_t *fbp = fbdev_obj->act_buffer;
+          for (y = y1; y <= y2; y++)
+            {
+              FAR uint8_t *cur_pos = fbp + (y * xres) + x1;
+              for (x = x1; x <= x2; x++)
+                {
+                  *cur_pos = lv_color_to8(*color_p);
+                  cur_pos++;
+                  color_p++;
+                }
+            }
+        }
+        break;
+      default:
+        break;
+    }
+
+  LV_LOG_TRACE("end copy");
+
+  fbdev_update_part(fbdev_obj, disp_drv, area_p);
+}
+
+/****************************************************************************
+ * Name: fbdev_get_pinfo
+ ****************************************************************************/
+
+static int fbdev_get_pinfo(int fd, FAR struct fb_planeinfo_s *pinfo)
+{
+  int ret = ioctl(fd, FBIOGET_PLANEINFO,
+                  (unsigned long)((uintptr_t)pinfo));
+  if (ret < 0)
+    {
+      LV_LOG_ERROR("ERROR: ioctl(FBIOGET_PLANEINFO) failed: %d", errno);
+      return ret;
+    }
+
+  LV_LOG_INFO("PlaneInfo (plane %d):", pinfo->display);
+  LV_LOG_INFO("    fbmem: %p", pinfo->fbmem);
+  LV_LOG_INFO("    fblen: %lu", (unsigned long)pinfo->fblen);
+  LV_LOG_INFO("   stride: %u", pinfo->stride);
+  LV_LOG_INFO("  display: %u", pinfo->display);
+  LV_LOG_INFO("      bpp: %u", pinfo->bpp);
+
+  /* Only these pixel depths are supported.  viinfo.fmt is ignored, only
+   * certain color formats are supported.
+   */
+
+  if (pinfo->bpp != 32 && pinfo->bpp != 16 &&
+      pinfo->bpp != 8  && pinfo->bpp != 1)
+    {
+      LV_LOG_ERROR("bpp = %u not supported", pinfo->bpp);
+      return -1;
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: fbdev_try_init_fbmem2
+ ****************************************************************************/
+
+static int fbdev_try_init_fbmem2(FAR struct fbdev_obj_s *state)
+{
+  uintptr_t buf_offset;
+  struct fb_planeinfo_s pinfo;
+
+  memset(&pinfo, 0, sizeof(pinfo));
+
+  /* Get display[1] planeinfo */
+
+  pinfo.display = state->pinfo.display + 1;
+
+  if (fbdev_get_pinfo(state->fd, &pinfo) < 0)
+    {
+      return -1;
+    }
+
+  /* check displey and match bpp */
+
+  if (!(pinfo.display != state->pinfo.display
+     && pinfo.bpp == state->pinfo.bpp))
+    {
+      LV_LOG_WARN("fbmem2 is incorrect");
```

Comment:
There are two ways to allocate double-frame framebuffers, one is traditional two-frame buffers next to consecutive addresses, relying on `pinfo.yoffset` to switch the buffer to be displayed. However, on embedded devices, the size of a single block of SRAM may be limited and cannot accommodate two frames of buffers with consecutive addresses. It is necessary to place the two buffers on two discontinuous addresses. Here we use the method of the `fbdev_try_init_fbmem2` function, which will Re-acquire the address of the second buffer from the driver, so that the logic of the switch buffer (by setting `pinfo.yoffset`) can be shared perfectly. (this solution comes from @xiaoxiang781216 )

Reply:
> I'm still getting this warning using the `ttgo_t_display_esp32:lvgl_fb`. I couldn't figure out the implementation, but I'd say to double-check the examples that use framebuffer.
> 
> ```
> [LVGL] [Warn]   (18.030, +18030)         fbdev_try_init_fbmem2: fbmem2 is incorrect     (in lv_port_fbdev.c line #522)
> ```

I have lowered this LOG level to INFO level and can ignore this warning.